src/snowflake/mod.rs

```diff
@@ -92,6 +105,93 @@ impl object_store::CredentialProvider for S3StageCredentialProvider {
 }
 
 
+#[derive(Debug)]
+pub(crate) struct SnowflakeAzureExtension {
+    stage: String,
+    client: Arc<SnowflakeClient>,
+}
+
+#[async_trait::async_trait]
+impl Extension for SnowflakeAzureExtension {
+    fn as_any(&self) -> &dyn std::any::Any {
+        self
+    }
+    async fn current_stage_info(&self) -> crate::Result<String> {
+        let stage_info = &self
+            .client
+            .current_upload_info(&self.stage)
+            .await?
+            .stage_info;
+        let stage_info: NormalizedStageInfo = stage_info.try_into()?;
+        let string = serde_json::to_string(&stage_info)
+            .context("failed to encode stage_info as json").to_err()?;
+        Ok(string)
+    }
+}
+
+#[derive(Debug)]
+pub(crate) struct AzureStageCredentialProvider {
+    stage: String,
+    client: Arc<SnowflakeClient>,
+    cached: Mutex<Option<Arc<object_store::azure::AzureCredential>>>
+}
+
+impl AzureStageCredentialProvider {
+    pub(crate) fn new(stage: impl AsRef<str>, client: Arc<SnowflakeClient>) -> AzureStageCredentialProvider {
+        AzureStageCredentialProvider { stage: stage.as_ref().to_string(), client, cached: Mutex::new(None) }
+    }
+}
+
+#[async_trait::async_trait]
+impl object_store::CredentialProvider for AzureStageCredentialProvider {
+    type Credential = object_store::azure::AzureCredential;
+    async fn get_credential(&self) -> object_store::Result<Arc<Self::Credential>> {
+        let info = self.client.current_upload_info(&self.stage).await
+            .map_err(|e| {
+                object_store::Error::Generic {
+                    store: "MicrosoftAzure",
+                    source: e.into()
+                }
+            })?;
+
+        if info.stage_info.location_type != "AZURE" {
+            return Err(object_store::Error::Generic {
+                store: "MicrosoftAzure",
+                source: Error::invalid_response("Location type must be AZURE for this provider").into()
+            })
+        }
+
+        let new_creds = info.stage_info.creds.as_azure()
+            .map_err(|e| object_store::Error::Generic {
+                store: "MicrosoftAzure",
+                source: e.into()
+            })?;
+        
+
+        let token_bytes = new_creds.azure_sas_token.trim_start_matches('?').as_bytes();
+        let new_pairs = url::form_urlencoded::parse(token_bytes)
+            .into_owned()
+            .collect();
+        
+        let mut locked = self.cached.lock().await;
+
+        match locked.as_ref() {
+            Some(creds) => {
+                if matches!(creds.as_ref(), AzureCredential::SASToken(pairs) if *pairs == new_pairs) {
+                    return Ok(Arc::clone(creds));
+                }
+            }
+            _ => {}
+        }
```

Comment:
@andrebsguedes: I adapted this from the S3 code + your variant, but don't follow why we have this caching here. 

Speaking about the S3 variant: the request to Snowflake in `current_upload_info` (which happens before the cache check) should be more expensive than  creating an `object_store::aws::AwsCredential` (which happens after), right? I also don't have a good understanding of when `get_credential` is called, maybe that is my problem.

Reply:
Yeah, you are right that this caching here is very confusing and my Azure version of this was pretty dumb to be honest, let me try to explain the rationale:
- object_store calls this for every operation, even multiple times in some cases, so we want this to be fast
- `current_upload_info` is a cached call to `fetch_upload_info` which means that most of the time calling it should just be a few atomic operations (to read the upload_info cache and perform the `Arc::clone`) and thus fast.
- So we call `current_upload_info` but instead of cloning all the strings to build `AwsCredential` we cache it so that we can `Arc::clone` it most of the time

Some critical analysis after looking at my S3 code again: 
- In reality even if everything goes great the extra contention from the lock may be way worse than the allocations for constructing `AwsCredential`.
- We would have to employ way more clever tricks with `arc_swap` and use the address of the `current_upload_info` result to detect change instead of checking the contents

TLDR: I think this caching is not worth it neither in S3 or Azure and we should drop it for now

Reply:
Agree to drop it. Let's do that in another PR. So I will keep this as is here